### ISSUE TITLE
Every call to wavesurferjs keeps on increasing RAM memory, without releasing  it on destroy.

### ISSUE THREAD

**Issue by EABangalore** [katspaugh/wavesurfer.js#1940]
I have came across a issue with `wavesurferjs` it does not release memory previously held upon destroy.

With high duration audio file, if we open 10 times it will hang browser itself (then we need to either hard refresh or close browser itself to continue).

Steps to reproduce on linux:

1.open terminal type $htop
2.keep on opening audio with destroy on closing.
3.memory will keep on adding

Though i'm not a good programmer, but i have found a exact solution with `canvas`  and `audio api`

Here is that solution: https://stackoverflow.com/questions/53241345/web-audio-api-memory-leak?noredirect=1&lq=1

i request @katspaugh , please implement it as soon as possible.

**Comment by marizuccara** [comment]
Hi, which backend are you using? `WebAudio` or `MediaElementWebAudio`?

**Comment by EABangalore** [comment]
Hi @marizuccara , i have not provided `backend` option itself while initialization. that means i'm using `webAudio`.

**Comment by marizuccara** [comment]
`WebAudio` is not recommended for high duration audio files, because it decodes internally your file audio to obtain the data necessary to draw the waveform. But with huge audio files this operation is too expensive, because it stores the data in memory. This can lead to browser crashes.

From the `WebAudio API` documentation:
![image](https://user-images.githubusercontent.com/44341274/81602742-717fc880-93cd-11ea-8354-7b6ca15c23a4.png)

Link: https://developer.mozilla.org/en-US/docs/Web/API/AudioBuffer

Another solution is to use `MediaElementWebAudio` backend which does not decode internally the audio data. This is because `MediaElementWebAudio` use the HTML5 audio tag, which performs buffering. However with this backend you have to pass the waveform peaks manually.

**Comment by thijstriemstra** [comment]
> Though i'm not a good programmer, but i have found a exact solution with canvas and audio api
> Here is that solution: https://stackoverflow.com/questions/53241345/web-audio-api-memory-leak?noredirect=1&lq=1

Have you tested it @EABangalore? if you have any fixes, please suggest them here and/or open a pull request!

**Comment by EABangalore** [comment]
@thijstriemstra , i don't  see closing of audioContext reference in `destroy()`

```
    destroy() {
        this.destroyAllPlugins();
        this.fireEvent('destroy');
        this.cancelAjax();
        this.clearTmpEvents();
        this.unAll();
        if (this.params.responsive !== false) {
            window.removeEventListener('resize', this._onResize, true);
            window.removeEventListener(
                'orientationchange',
                this._onResize,
                true
            );
        }
        if (this.backend) {
            this.backend.destroy();
        }
        if (this.drawer) {
            this.drawer.destroy();
        }
        this.isDestroyed = true;
        this.isReady = false;
        this.arraybuffer = null;
    }
```

As per this **mdn** reference we should be doing something like this https://developer.mozilla.org/en-US/docs/Web/API/AudioContext/close

```
var audioCtx = new AudioContext();
audioCtx.close().then(function() { ... });
await audioCtx.close();
```

**Comment by marizuccara** [comment]
The destroy of AudioContext is made in `this.backend.destroy()`. If you go to backend, that is `webaudio.js`, you can see the destroy method where are destroyed stuffs related to WebAudio

**Comment by ssp0929** [comment]
I've noticed that JSArrayBuffers just stay in memory even after calling .destroy() and GC only picks them up after a refresh.

This is kind of painful for SPAs that by nature do not refresh.

**Comment by marizuccara** [comment]
Hi, are you sure are you using WebAudio as backend? You are calling the destroy method somewhere?

**Comment by thijstriemstra** [comment]
Can you use `measureMemory()` in Chrome (83 and newer) to monitor the memory usage and report results here?

See https://web.dev/monitor-total-page-memory-usage/

**Comment by marcelblum** [comment]
Hello, I've been noticing this same issue on my end (Win 10, latest Chrome, latest Wavesurfer). I can confirm that Wavesurfer is erroneously not releasing the audio buffer after a call to `destroy()`, which results in the buffer not being garbage collected and persisting in memory, even after subsequent calls to `load()` different sound files using the same Wavesurfer instance. This is most apparent when using large audio files. I know those MDN docs recommend short sound files but in practice on a decent machine Chrome handles larger audio files beautifully and I often work with tracks >1 hour in length as decoded buffers with the Web Audio API with no issues. If you set a variable holding an audio buffer to `null` the memory is typically purged within seconds by the garbage collector. Therefore this is almost certainly not a Web Audio API/browser bug. Wavesurfer handles drawing large sounds beautifully too, but even after a call to `destroy` I can see that there are still retained references to the buffer in the Wavesurfer object, preventing garbage collection. Simple example, go to https://wavesurfer-js.org/ and open the console:

```
wavesurfer.destroy();
console.log(wavesurfer.backend.source.buffer);
//->AudioBuffer {length: 1045146, duration: 21.773875, sampleRate: 48000, numberOfChannels: 1}
//buffer is still assigned to a variable and can't be garbage collected!
wavesurfer.backend.destroy();
console.log(wavesurfer.backend.source.buffer);
//->AudioBuffer {length: 1045146, duration: 21.773875, sampleRate: 48000, numberOfChannels: 1}
//even now buffer is still referenced in the wavesurfer object!
```

However even if I try manually `wavesurfer.backend.source.buffer = null` the buffer is still not garbage collected so I suspect there remains some reference to it even elsewhere in the Wavesurfer code that I haven't figured out yet, perhaps you have an idea?

And just as a demonstration here is simple vanilla Web Audio API code that loads a large audio file and then purges it when done, easily observable in Windows Task Manager:

```
var audioCtx = new AudioContext();
var buffer;
var request = new XMLHttpRequest();
request.open("GET", "https://soundplant.org/fruits/Christian%20Wolff%20-%20Edges%20-%20Gentle%20Fire.mp3");
request.responseType = "arraybuffer";
request.onload = function() {
    console.log("loaded");
    let undecodedAudio = request.response;
    audioCtx.decodeAudioData(undecodedAudio, (data) => {
        console.log("decoded");
        buffer = data;
    });
};
request.send();
/* Chrome loads ~300mb audio data into memory observable in Task Manager */
buffer = null;
/* usually within a few seconds, 300mb purged from Chrome's used memory observable in Task Manager */
```

**Comment by thijstriemstra** [comment]
thanks for the details @marcelblum. 

> However even if I try manually wavesurfer.backend.source.buffer = null the buffer is still not garbage collected so I suspect there remains some reference to it even elsewhere in the Wavesurfer code that I haven't figured out yet, perhaps you have an idea?

Can you compare the memory size of the vanilla example with wavesurfer.js? Wavesurfer draws many Canvas instances and they add up in memory as well. Would be useful to be able to tell how much WebAudio used and how much Canvas used.

**Comment by marcelblum** [comment]
yes I have compared and they are nearly identical. Even with all of Wavesurfer's DOM/Canvas stuff the amount of RAM used by that is negligible compared to 300MB of audio (like maybe 20mb). That's at least in my own usage, with no Wavesurfer plugins or interactivity, just using Wavesurfer for visualizing. Obviously there are some browser caching and memory management quirks out of our control and it doesn't always release memory when we want it to, but it usually does. Here are 2 test pages I set up as I am trying to determine the most reliable way to purge a large audio buffer from memory in order to make room for loading another without having to refresh the page:

https://soundplant.org/wavesurfer/vanillawebaudio.html
https://soundplant.org/wavesurfer/wavesurfer.html

In the vanilla web audio test, Chrome doesn't always purge the buffer instantaneously after setting the reference to null, but it's usually within seconds, sometimes a minute or 2. In the Wavesurfer test the only way I can get Chrome to purge is if I `load()` a small dummy file.

Interestingly, in a quick test in Win10/Firefox, `wavesurfer.backend.source.buffer = null` seems to do the trick and purge instantaneously. But in Chrome it never does.

In any case, in part this is running up against browser quirks but there is definitely a Wavesurfer bug here if `wavesurfer.destroy()` doesn't properly release its reference to the buffer.

**Comment by marcelblum** [comment]
BTW on further clarification I was wrong when I said

> the buffer not being garbage collected and persisting in memory even after subsequent calls to load() different sound files using the same Wavesurfer instance

It's clear that `load()` reuses the same variable to store the buffer and therefore overwrites the previously loaded one when reusing the same Wavesurfer instance. So for my needs when I want to purge memory I `load()` a small dummy file. But that's rather ugly.

**Comment by katspaugh** [comment]
Thank you Marcel, you nailed it! Let’s clear that buffer.
I think it might be also trapped in some closure (an event callback?). We should look for places where the buffer is passed as an argument or assigned to a local variable.

**Comment by marizuccara** [comment]
Hi @thijstriemstra have you got a suggestion of what to fix? Check if AudioBuffer is really empty after calling the destroy method?

**Comment by a2-rmiller** [comment]
AudioBuffer did not decrease in size and references to the buffers was being kept around even after all instances were destroyed after testing using multiple heap snapshots in Chrome.

On my end, I resolved this (before doing anything) by straight up deleting the buffer first.

delete this.wave.backend.buffer;

After the buffer is deleted, I run my regular destroying things like this.wave.destroy(); etc.

This solves the memory leak for me, which was upwards of 400mb every instantiation of a new object (using any of the recoemmdned backends).

**Comment by EABangalore** [comment]
@a2-rmiller, thanks for your reply 

I have tried your solution, but it did not help me i'm using `wavesurfer.js 3.3.3`

here is what i have tried

```
var wavesurfer = {}; // defined globally
wavesurfer = WaveSurfer.create({
          backend: 'MediaElementWebAudio',
         container: '#waveform',
       ...
});
wavesurfer.load(audioUrl,[],'none');

getAudioPeakFromJSON(function(peakData){
            var audio = document.createElement('audio');
            audio.src = audioUrl;
            // Set crossOrigin to anonymous to avoid CORS restrictions
            audio.crossOrigin = 'anonymous';
             wavesurfer.load(audio,peakData,'auto');
  });

//destroying this way

delete wavesurfer.wave.backend.buffer;

delete wavesurfer.wave.destroy();
```
@a2-rmiller, what i'm doing wrong please suggest me

Thanks in advance!!

**Comment by a2-rmiller** [comment]
@EABangalore I'll list the things I am doing differently than you based on that small snippet:

- I am not calling destroy as soon as I create it, but laster on when I need it destroyed
- I am using version 4.0.1
- My audio files are wavs
- I am not using the extra parameters in the wavefurfer.load
- I am not using a function regarding peakdata
- I use .unAll(); before .destroy() (since I use wavesurfer events)
- My instance of wavesurfer is not defined globally (in the same way), but part of a class structure, where instantiation is called inside of a function to easily tear down and recreate everything

I hope this helps you.

**Comment by thijstriemstra** [comment]
@a2-rmiller can you turn this potential fix into a pull request?

**Comment by a2-rmiller** [comment]
@thijstriemstra Sorry I have not dug into the actual codebase and would not directly have the time to do so. However, destroying the instantiated AudioBuffer before destroying other things would be the route to go (wherever this may live in the source).

**Comment by a2-rmiller** [comment]
@thijstriemstra If I had to guess though:
https://github.com/katspaugh/wavesurfer.js/blob/7af166d97009afd9e6b98726968dd0af3e1126b1/src/wavesurfer.js#L1690
Adding a check here to truly check if those items are destroyed before settings destroyed to true would likely reveal something more.

**Comment by sundayz** [comment]
Hi guys, gonna share my findings on this issue. I'm testing with 2 instances of wavesurfer per page.

When not using pre-loaded peaks and using WebAudio backend:
![image](https://user-images.githubusercontent.com/11685806/88486374-12196980-cf75-11ea-857f-f65dce8d6337.png)

1st memory snapshot: after app loads.
2nd memory snapshot: after 2 tracks are deleted (destory method is called, DOM node is deleted)

In this case the largest memory use comes from two ArrayBuffers, 79MB which corresponds to the size of backend.buffer for track 1 + track 2 

When using preloaded peaks and using WebAudioMediaElement backend:
- 28MB used after load
- 25MB used after destroy

Calling .unAll before destroy() makes no difference, and it shouldn't, since it looks like destroy() calls unAll as well.

Using `delete wavesurfer.backend.buffer` makes no difference. The memory is only freed after the browser tab is closed.

**Comment by thijstriemstra** [comment]
> Using delete wavesurfer.backend.buffer makes no difference.

Was afraid so.

**Comment by sundayz** [comment]
it turns out that Google Chrome has really good memory tools, it can tell you where things were allocated, what's holding on to the references, etc. 

I'm not that good at using it though, and I'm using 3.3.3 and Vue so it's harder for me to figure out what's going on. Someone who knows how to use the chrome memory tools could probably figure this out easily.

this is the allocation stack for the backing buffer that's not being GC'd when calling destroy()
![image](https://user-images.githubusercontent.com/11685806/88490181-9e398a00-cf91-11ea-8092-71c46ad83fe2.png)

The anonymous functions lead to decodeArrayBuffer -> offlineAc.decodeAudioData(). There's a lot of callbacks here and events, maybe a hint for someone who can work this out

**Comment by a2-rmiller** [comment]
This is unfortunately incorrect. Chrome does in fact have very nice memory snapshotting tools. When removing the delete buffer here is the snapshots:


Heaps:
![image](https://user-images.githubusercontent.com/65678254/88543378-7839d700-cfdd-11ea-843c-2e67f485c91c.png)



First Heap:

![image](https://user-images.githubusercontent.com/65678254/88543543-c2bb5380-cfdd-11ea-9579-fed0f5e9afac.png)


From the image above, this is the regular first time load where the buffer gets populated. 


Second Time:
![image](https://user-images.githubusercontent.com/65678254/88543686-f1d1c500-cfdd-11ea-9dcd-94f4ce5d6767.png)


Third Time:
![image](https://user-images.githubusercontent.com/65678254/88543738-04e49500-cfde-11ea-9637-e9ee816dba8d.png)


Fourth Time:
![image](https://user-images.githubusercontent.com/65678254/88543781-19289200-cfde-11ea-8c8e-18a2aa5a936d.png)



When clicking the instances of the bufferdata, it all leads back to:
![image](https://user-images.githubusercontent.com/65678254/88543915-4412e600-cfde-11ea-8ff0-3a0ccea83e86.png)

**Comment by sundayz** [comment]
can we pin this issue?

Also note to anyone who has issues with this, using the MediaElementWebAudio backend seems to reduce ram usage a LOT so you could use that for now

**Comment by thijstriemstra** [comment]
> can we pin this issue?

Done.

**Comment by jordanfloyd** [comment]
Just to add to the reports, was also seeing this issue in 4.0.1. Calling `delete wavesurfer.backend.buffer` does indeed release the memory for me.

**Comment by SeanBannister** [comment]
For anyone else having trouble diagnosing this. `delete wavesurfer.backend.buffer` seems to work but it's inconsistent. Running it 20 times in a row I would often get one time where it wouldn't work.

Also it only worked on MediaElementWebAudio it didn't work on WebAudio. I was testing this without providing Peak data as that increases memory usage.

Due to the inconsistencies (and me getting frustrated debugging this) here's a few potential tips when diagnosing this:
- It can take up to 5 seconds for the memory to be released so don't come to conclusions to quickly.
- Try opening a new tab every time you test. Refreshing the tab doesn't always release the memory.
- (Unsure about this) Try closing Chrome's developer tools. This seems very odd but it seems having developer tools open sometimes caused the memory to not be released. 🤷‍♂️
- Due to the above don't run `delete wavesurfer.backend.buffer` inside of Chrome's console put it in your own code.

I have no experience diagnosing memory leaks but found it odd that when I [turned on displaying "Javascript memory"](https://developers.google.com/web/tools/chrome-devtools/memory-problems#monitor_memory_use_in_realtime_with_the_chrome_task_manager) inside Chrome's Task Manager it isn't showing up as "Javascript memory" usage just as general "Memory footprint".

**Comment by SeanBannister** [comment]
Note that this memory leak affects more than `.destroy()` it's also an issue with memory not being released during general usage.

For example if you use `MediaElementWebAudio` and don't provide peak data you can run `delete wavesurfer.backend.buffer` after the Peak data is generated and it frees up a heap of memory.

Please note I don't understand the inner workings of this.

**Comment by entonbiba** [comment]
hmm for me the following happens with version 4.4.0 and chrome version 88.0.4324.87

- `.create()` and `.load()` clears the memory
- `.destroy` `backend.buffer = null` `.unAll` doesn't

![image](https://user-images.githubusercontent.com/5193884/105431659-0b1d9c00-5c24-11eb-96bd-380efbbfe172.png)

**Comment by katspaugh** [comment]
Great find! What is it that `.load()` does clears the memory?

**Comment by thijstriemstra** [comment]
> We will need to add wavesurfer.backend = null; on destroy() method as this will release usage.

Sounds good.

**Comment by entonbiba** [comment]
@thijstriemstra patched here #2164

**Comment by thijstriemstra** [comment]
can everyone participating in this ticket give #2164 a try?
